posthog/api/query.py

```diff
@@ -0,0 +1,82 @@
+import json
+from typing import Dict
+
+from django.http import HttpResponse, JsonResponse
+from drf_spectacular.types import OpenApiTypes
+from drf_spectacular.utils import OpenApiParameter
+from rest_framework import viewsets
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.request import Request
+
+from posthog.api.documentation import extend_schema
+from posthog.api.routing import StructuredViewSetMixin
+from posthog.exceptions import RequestParsingError
+from posthog.models import Team
+from posthog.models.event.query_event_list import query_events_list_v2
+from posthog.permissions import ProjectMembershipNecessaryPermissions, TeamMemberAccessPermission
+from posthog.rate_limit import PassThroughClickHouseBurstRateThrottle, PassThroughClickHouseSustainedRateThrottle
+from posthog.schema import EventsQuery
+
+
+class QueryViewSet(StructuredViewSetMixin, viewsets.ViewSet):
+    permission_classes = [IsAuthenticated, ProjectMembershipNecessaryPermissions, TeamMemberAccessPermission]
+    throttle_classes = [PassThroughClickHouseBurstRateThrottle, PassThroughClickHouseSustainedRateThrottle]
+
+    @extend_schema(
+        parameters=[
+            OpenApiParameter(
+                "query",
+                OpenApiTypes.STR,
+                description="Query node JSON string",
+            ),
+        ]
+    )
+    def list(self, request: Request, **kw) -> HttpResponse:
+        query_json = self._query_json_from_request(request)
+        query_result = process_query(self.team, query_json)
+        return JsonResponse(query_result)
+
+    def post(self, request, *args, **kwargs):
+        query_json = request.data
+        query_result = process_query(self.team, query_json)
+        return JsonResponse(query_result)
+
+    def _query_json_from_request(self, request):
+        if request.method == "POST":
+            if request.content_type in ["", "text/plain", "application/json"]:
+                query_source = request.body
+            else:
+                query_source = request.POST.get("query")
+        else:
+            query_source = request.GET.get("query")
+
+        if query_source is None:
+            raise RequestParsingError("Please provide a query in the request body or as a query parameter.")
+
+        try:
+            # parse_constant gets called in case of NaN, Infinity etc
+            # default behaviour is to put those into the DB directly
```

Comment:
Looks like the "put those into the DB directly" comment is neatly copied from capture endpoint code. :D It doesn't seem really correct to allow such requests though, as they'll result in "undefined behavior".

Reply:
makes sense. will raise.